shared-bindings/canio/Listener.h

```diff
@@ -27,5 +27,17 @@
 #pragma once
 
 #include "py/obj.h"
+#include "shared-bindings/canio/CAN.h"
+#include "shared-bindings/canio/Match.h"
 
 extern const mp_obj_type_t canio_listener_type;
+
+typedef struct canio_listener_obj canio_listener_obj_t;
+
+void common_hal_canio_listener_construct(canio_listener_obj_t *self, canio_can_obj_t *can, size_t nmatch, canio_match_obj_t **matches, float timeout);
+void common_hal_canio_listener_check_for_deinit(canio_listener_obj_t *self);
+void common_hal_canio_listener_deinit(canio_listener_obj_t *self);
+bool common_hal_canio_listener_receiveinto(canio_listener_obj_t *self, canio_message_obj_t *message);
```

Comment:
I'd just have this as `receive` and return an `mp_obj_t`. That way you can have two separate structure for Message and RTR.

Reply:
That won't work out, because the type of message must be determined by the hardware.

Reply:
Let me amplify on a hastily written answer.

When calling read() or __next__(), the software **has** to commit to taking the next messge that has arrived on this FIFO, regardless of whether it is a RTR or not.  This is true of all 3 HW implementations I've looked at (SAM E, STM, and ESP32S2).

Having the common_hal routine read into a single structure which can accommodate either type of message is the simplest way I see to do it.  Adding a "peek" routine, a union type, or other complication is a needless complication, compared to having an "rtr" field in the C struct.

This implementation detail is transparent to the user, because the two types become distinct Python types with distinct behaviors (one has a data property, one has a size property).  As we've discussed in other contexts, the 8 bytes of storage set aside for the message that have to be allocated regardless of whether it's an RTR are also not going to break the bank, it's still within that single GC heap block.

I can change it if it's important, but I don't think it'll be an improvement.  If you do want me to change it, please show me what the common_hal APIs should be called and what their parameter lists should look like so that I deliver something that is satisfactory.

Reply:
I went ahead and wrote what I think you want.  It adds about 300 lines of largely duplicate code, because
 * now common hal's implementation listener's receive needs two copies of the code, to read the HW registers into two different structures
 * and an additional common hal send remote transmission request API is now needed, which duplicates copying the new rtr structure into the hw registers

This also amounts to about 300 extra bytes of firmware, but for me it is the duplicated code that is the most objectionable.

e.g., this had to be added, while in the existing function RTR is hard-coded as false instead of as message->rtr:
```
+void common_hal_canio_can_send_remote_transmission_request(canio_can_obj_t *self, canio_remote_transmission_request_obj_t *remote_transmissi
on_request)
+{
+    maybe_auto_restart(self);
+
+    // We have just one dedicated TX buffer, use it!
+    canio_can_tx_buffer_t *ent = &self->state->tx_buffer[0];
+
+    ent->txb0.bit.ESI = false;
+    ent->txb0.bit.XTD = remote_transmission_request->extended;
+    ent->txb0.bit.RTR = true;
+    if (remote_transmission_request->extended) {
+        ent->txb0.bit.ID = remote_transmission_request->id;
+    } else {
+        ent->txb0.bit.ID = remote_transmission_request->id << 18; // short addresses are left-justified
+    }
+
+    ent->txb1.bit.MM = 0; // "remote_transmission_request marker"
+    ent->txb1.bit.EFC = 0; // don't store fifo events to event queue
+    ent->txb1.bit.FDF = 0; // Classic CAN format
+    ent->txb1.bit.BRS = 0; // No bit rate switching
+    ent->txb1.bit.DLC = remote_transmission_request->size;
+
+    // TX buffer add request
+    self->hw->TXBAR.reg = 1;
+
+    // wait 8ms (hard coded for now) for TX to occur
+    uint64_t deadline = port_get_raw_ticks(NULL) + 8;
+    while (port_get_raw_ticks(NULL) < deadline && !(self->hw->TXBTO.reg & 1)) {
+        RUN_BACKGROUND_TASKS;
+    }
+}
+
```

changed/added internal APIs are:
```
-bool common_hal_canio_listener_receiveinto(canio_listener_obj_t *self, canio_message_obj_t *message);
+mp_obj_t common_hal_canio_listener_receive(canio_listener_obj_t *self);
```
and
```
-void common_hal_canio_can_send(canio_can_obj_t *self, canio_message_obj_t *message);
+void common_hal_canio_can_send_message(canio_can_obj_t *self, canio_message_obj_t *message);
+void common_hal_canio_can_send_remote_transmission_request(canio_can_obj_t *self, canio_remote_transmission_request_obj_t *remote_transmission_request);
```
and
```
-bool common_hal_canio_message_get_rtr(const canio_message_obj_t *self);
-void common_hal_canio_message_set_rtr(canio_message_obj_t *self, bool rtr);
-void common_hal_canio_remote_transmission_request_set_length(canio_message_obj_t *self, size_t length);
```

Reply:
The comment "+    ent->txb1.bit.MM = 0; // "remote_transmission_request marker"" is just wrong btw